onnxruntime/core/framework/execution_frame.h

```diff
@@ -168,5 +177,11 @@ class ExecutionFrame final : public IExecutionFrame {
 
   // Big chunks on different locations that will be used by mem_pattern.
   std::map<OrtMemoryInfo, BufferUniquePtr> buffers_;
+
+  // Given the input shapes of the executed graph, ExecutionFrame tries inferring
+  // all symbolic shapes. inferred_shapes_.map_[i] is the shape of OrtValue indexed
+  // by i, if the key i exists.
+  // inferred_shapes_ is generated togehter with mem_patterns_.
+  ThreadSafeUnorderedMap<int, TensorShape> inferred_shapes_;
```

Comment:
>ThreadSafeUnorderedMap [](start = 2, length = 22)

why this needs to be thread safe? Execution frame usually is single thread

Reply:
ParallelExecutor. A Windows test was failed because we use thread-unsafe map to track memory usage (https://github.com/microsoft/onnxruntime/pull/4454). With this thread-safe map, we can revert that PR and bring memory tracker back.

Reply:
Is this test related to Send/Recv revision? If not maybe we should separate into a different PR.

---
In reply to: [462616811](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r462616811) [](ancestors = 462616811)

Reply:
The current parallel executor requires ExecutionFrame to be thread-safe (this requirement was discovered by our memory tracker). Now, memory tracker is removed but whatever we introduce into ExecutionFrame, it needs to be thread-safe.

---
In reply to: [462619306](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r462619306) [](ancestors = 462619306,462616811)

Reply:
Thread safety could be more than in member variables. There are relationships between different data structures in this class, and we need to make the function interface thread-safe. Otherwise, even if inferred_shapes_ is thread-safe, it could contain inconsistent info to other member variables.

---
In reply to: [462622259](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r462622259) [](ancestors = 462622259,462619306,462616811)

Reply:
Yes, but making entire ExecutionFrame thread-safe is beyond the scope of this PR.

---
In reply to: [462626332](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r462626332) [](ancestors = 462626332,462622259,462619306,462616811)

Reply:
Yes that's why I felt anything related to thread-safe should be in a separate PR

---
In reply to: [462653034](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r462653034) [](ancestors = 462653034,462626332,462622259,462619306,462616811)

Reply:
We can't break master branch.. Are we going to revise parallel executor before checking this PR?

---
In reply to: [462655789](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r462655789) [](ancestors = 462655789,462653034,462626332,462622259,462619306,462616811)